src/graphql.js

```diff
@@ -61,42 +54,229 @@ itemToString = ({name, args, children}) => {
   * @method toString - Prints out the query as a string. Required by the runtime
   * for query execution, also handy for debugging.
   */
-const queryNode = (name, args = {}, children = []) => {
-  const item = {name, args, children}
+const queryRoot = ({name, args, children, type}) => {
+  const item = {name, args, children, type}
 
-  item.addChild = child => {
-    return queryNode(name, args, children.concat(child))
-  }
+  item.addChild = child =>
+    queryRoot({name, args, children: children.concat(child), type})
 
-  item.toString = () => itemToString(item)
+  item.toString = () => {
+    if (args == null) {
+      throw new Error(`No args passed to ${name}`)
+    }
+    return name + argsString(args) + childrenString(children)
+  }
 
   return item
 }
 
+const queryLeaf = name => queryRoot({
+  name,
+  args: {},
+  children: [],
+  type: 'leaf'
+})
+
+const queryNoid = (name, args, children) => queryRoot({
+  name,
+  args,
+  children,
+  type: 'noid'
+})
+
+const queryNode = (name, args, children) => queryRoot({
+  name,
+  args,
+  children: children.concat([queryLeaf('id'), queryLeaf('__typename')]),
+  type: 'node'
+})
+
+const pagination = queryRoot({
+  name: 'pageInfo',
+  args: {},
+  children: [queryLeaf('endCursor'), queryLeaf('hasNextPage')],
+  type: 'pagination meta'
+})
+
+const queryEdge = (name, args, children) => {
+  args.first = args.first || 1
+
+  return queryRoot({
+    name,
+    args,
+    type: 'edge',
+    children: [pagination]
+  }).addChild(queryNode('nodes', {}, children))
+}
+
+const queryOn = (type, children) =>
+      queryNode(`... on ${type}`, {}, children)
+
+const queryType = (name, args, types) => {
+  const children = types.map(([type, children]) => queryOn(type, children))
+  return queryRoot({
+    name,
+    args,
+    type: 'typed',
+    children: children
+  })
+}
+
+// -----
+// Query Depagination
+// -----
+
+// Kludgy non-destructive assoc
+const assoc = (o, k, v, k2, v2) => {
+  const out = {}
+  Object.assign(out, o)
+  out[k] = v
+  if (k2 && v2) {
+    out[k2] = v2
+  }
+  return out
+}
+
+const find = (l, k) => {
```

Comment:
Can we have better comments on when these individual functions are used?

Reply:
We can. I'll go through and add some.

Reply:
Thanks.